medusa/storage/google_storage.py

```diff
@@ -13,106 +13,286 @@
+        loop.run_until_complete(self._delete_objects(objects))
 
-        # ensure srcs is always a list
-        if isinstance(srcs, str) or isinstance(srcs, Path):
-            srcs = [srcs]
+    async def _delete_objects(self, objects: t.List[AbstractBlob]):
+        coros = [self._delete_object(obj) for obj in objects]
+        await asyncio.gather(*coros)
 
-        generators = self._download_blobs(srcs, dest)
-        return list(itertools.chain(*generators))
+    async def _delete_object(self, obj: AbstractBlob):
+        await self.gcs_storage.delete(
+            bucket=self.bucket_name,
+            object_name=obj.name,
+            timeout=-1,
+        )
 
-    def _download_blobs(self, srcs, dest):
-        with GSUtil(self.config) as gsutil:
-            for parent, src_paths in _group_by_parent(srcs):
-                yield self._download_paths(gsutil, parent, src_paths, dest)
+    def download_blobs(self, srcs: t.List[t.Union[Path, str]], dest: t.Union[Path, str]):
+        loop = self._get_or_create_event_loop()
+        loop.run_until_complete(self._download_blobs(srcs, dest))
+
+    async def _download_blobs(self, srcs: t.List[t.Union[Path, str]], dest: t.Union[Path, str]):
+        coros = [self._download_blob(src, dest) for src in map(str, srcs)]
+        await asyncio.gather(*coros)
+
+    @retry(stop_max_attempt_number=MAX_UP_DOWN_LOAD_RETRIES, wait_fixed=5000)
+    async def _download_blob(self, src: str, dest: str):
+        blob = await self._stat_blob(src)
+        object_key = blob.name
+
+        # we must make sure the blob gets stored under sub-folder (if there is any)
+        # the dest variable only points to the table folder, so we need to add the sub-folder
+        src_path = Path(src)
+        file_path = (
+            "{}/{}/{}".format(dest, src_path.parent.name, src_path.name)
+            if src_path.parent.name.startswith(".")
+            else "{}/{}".format(dest, src_path.name)
+        )
 
-    def _download_paths(self, gsutil, parent, src_paths, old_dest):
-        new_dest = '{}/{}'.format(old_dest, parent) if parent.startswith('.') else old_dest
-        # we made src_paths a list of Path objects, but we need strings for copying
-        # plus, we must not forget to point them to the bucket
-        srcs = ['gs://{}/{}'.format(self.bucket.name, str(p)) for p in src_paths]
-        return gsutil.cp(
-            srcs=srcs,
-            dst=new_dest,
-            parallel_process_count=self.config.concurrent_transfers)
+        logging.debug(
+            '[Storage] Downloading gcs://{}/{} -> {}'.format(
+                self.config.bucket_name, object_key, file_path
+            )
+        )
 
-    def get_object_datetime(self, blob):
-        logging.debug("Blob {} last modification time is {}".format(blob.name, blob.extra["last_modified"]))
-        return parser.parse(blob.extra["last_modified"])
+        try:
+            await self.gcs_storage.download_to_filename(
```

Comment:
**Issue:** This method reads the files at once and puts their content into memory. We need to use [download_stream()](https://github.com/talkiq/gcloud-aio/blob/b170cf916f0ad52357bd3457ce5b905dd9170132/storage/gcloud/aio/storage/storage.py#L352C1-L352C1) instead which will return a BufferedStream which should be writeable to a file without loading all the data in memory.

Reply:
Done. Pushed a new commit, together with a rebase on a recent master.